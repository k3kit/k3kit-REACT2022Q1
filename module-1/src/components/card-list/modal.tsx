import React, { Component } from 'react';

interface a {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: {
    name: string;
    url: string;
  };
  location: {
    name: string;
    url: string;
  };
  image: string;
  episode: [];
  url: string;
  created: string;
}

interface MyProps {
  it: a;
  active: boolean;
}

export class Modal extends Component<MyProps> {
  render() {
    return <div className="modal"></div>;
  }
}
