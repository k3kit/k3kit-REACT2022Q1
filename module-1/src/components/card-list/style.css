.card-item{
    display: flex;
    flex-wrap: wrap;
    gap: 20px;
    justify-content: flex-start;
    width: 100%;
    max-width: 1600px;
    margin: 0 auto;
    padding-bottom: 40px;
}