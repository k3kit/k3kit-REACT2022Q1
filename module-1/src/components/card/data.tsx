const data = [
  {
    id: 1,
    avatar: 'https://robohash.org/culpamagnameius.png?size=50x50&set=set1',
    first_name: 'Ignaz',
    email: 'ibrasted0@arizona.edu',
    gender: 'Male',
  },
  {
    id: 2,
    avatar: 'https://robohash.org/saepelaboriosamofficia.png?size=50x50&set=set1',
    first_name: 'Raoul',
    email: 'rfissenden1@google.ru',
    gender: 'Female',
  },
  {
    id: 3,
    avatar: 'https://robohash.org/autinest.png?size=50x50&set=set1',
    first_name: 'Mavis',
    email: 'msimcock2@simplemachines.org',
    gender: 'Male',
  },
  {
    id: 4,
    avatar: 'https://robohash.org/asperiorestemporaincidunt.png?size=50x50&set=set1',
    first_name: 'Ave',
    email: 'abovingdon3@tmall.com',
    gender: 'Male',
  },
  {
    id: 5,
    avatar: 'https://robohash.org/velitporroid.png?size=50x50&set=set1',
    first_name: 'Eolande',
    email: 'efrascone4@wufoo.com',
    gender: 'Male',
  },
  {
    id: 6,
    avatar: 'https://robohash.org/debitisomnisnesciunt.png?size=50x50&set=set1',
    first_name: 'Arlie',
    email: 'alansly5@360.cn',
    gender: 'Male',
  },
  {
    id: 7,
    avatar: 'https://robohash.org/temporaetmaiores.png?size=50x50&set=set1',
    first_name: 'Rodie',
    email: 'rdoyley6@hubpages.com',
    gender: 'Male',
  },
  {
    id: 8,
    avatar: 'https://robohash.org/quiaquidemmagni.png?size=50x50&set=set1',
    first_name: 'Tymon',
    email: 'tmasterman7@scribd.com',
    gender: 'Male',
  },
];
export default data;
