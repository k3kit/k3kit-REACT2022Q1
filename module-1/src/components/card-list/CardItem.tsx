import { loadavg } from 'os';
import React, { Component } from 'react';
import Spinner from '../lodading/spinner';
import { Modal } from './modal';
import './style.css';
interface a {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: {
    name: string;
    url: string;
  };
  location: {
    name: string;
    url: string;
  };
  image: string;
  episode: [];
  url: string;
  created: string;
}

interface MyProps {
  el: a;
  loader: boolean;
}

export class CardItem extends Component<MyProps> {
  state = {
    modal: true,
  };

  handleActive = () => {
    this.setState({ modal: false });
  };
  render() {
    const { el } = this.props;
    let popup;
    if (this.props.loader) {
      return <Spinner />;
    }
    if (this.state.modal === false) {
      popup = (
        <div className="popup" onClick={() => this.setState({ modal: true })}>
          <div className="popup_inner" onClick={(e) => e.stopPropagation()}>
            <img className="card-img-popup" src={`${el.image}`} alt="avatar" />
            <div className="card-description-popup">
              <h3>Name: {el.name}</h3>
              <p>Status: {el.status}</p>
              <p>Species: {el.species}</p>
              <p>Gender: {el.gender}</p>
              <p>Created: {el.created}</p>
              <p>Origin: {el.origin.name}</p>
              <p>Location: {el.location.name}</p>
            </div>
          </div>
        </div>
      );
    }
    return (
      <>
        <li className="card" key={el.id} onClick={this.handleActive}>
          <img className="card-img" src={`${el.image}`} alt="avatar" />
          <div className="card-description">
            <h3>Name: {el.name}</h3>
            <p>Status: {el.status}</p>
            <p>Species: {el.species}</p>
          </div>
        </li>
        {popup}
      </>
    );
  }
}
