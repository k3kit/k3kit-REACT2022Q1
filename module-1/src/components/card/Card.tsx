import React, { FC } from 'react';
import './style.css';
interface Icard {
  id?: number;
  avatar: string;
  first_name: string;
  email: string;
  gender: string;
}

const Card: FC<Icard> = ({ avatar, first_name, email, gender }: Icard) => {
  return (
    <li className="card">
      <img className="card-img" src={`${avatar}`} alt="avatar" />
      <div className="card-description">
        <p className="first-name">Name:{first_name}</p>
        <p className="email">Email: {email}</p>
        <p className="gender">Gender: {gender}</p>
      </div>
    </li>
  );
};

export default Card;
