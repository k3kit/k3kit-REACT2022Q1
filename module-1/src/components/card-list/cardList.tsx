/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { FC, useContext, useEffect, useReducer, useState } from 'react';
import { Link } from 'react-router-dom';
import appContext from '../../context/app-context';
import Spinner from '../loading/spinner';
import { SearchBar } from '../search-bar/search';
import { CardItem, Icharacter } from './CardItem';
import { createPages } from './fun';
import './style.css';
// enum CardActionKind {
//   FETCH_SUCCESS = 'FETCH_SUCCESS',
//   FETCH_ERROR = 'FETCH_ERROR',
// }
// interface CardState {
//   isPending: boolean;
//   error: boolean;
//   character: Icharacter;
// }
// interface CardAction {
//   payload: Icharacter;
//   type: CardActionKind;
// }

export const CardList: FC = () => {
  const {
    value,
    currentPage,
    setCurrentPage,
    totalCount,
    isPending,
    error,
    character,
    FetchData,
    FetchError,
    FetchSlice,
    status,
    gender,
    species,
    setStatus,
    setGender,
    setSpecies,
  } = useContext(appContext);
  const spec = [
    'Human',
    'Alien',
    'Humanoid',
    'Poopybutthole',
    'Mythological',
    'Unknown',
    'Animal',
    'Disease',
    'Robot',
    'Cronenberg',
    'Planet',
  ];
  const url = `https://rickandmortyapi.com/api/character/?page=${currentPage}&name=${value}&status=${status}&gender=${gender}&species=${species}`;
  const genders = ['female', 'male', 'genderless', 'unknown'];
  const pagesCount = Math.ceil(totalCount / 20);
  const pages: any[] = [];
  createPages(pages, pagesCount, currentPage);
  useEffect(() => {
    const getData = async () => {
      try {
        const response = await fetch(url);
        if (response.status >= 400) {
          FetchError();
          setCurrentPage(1);
          console.log(url);
        } else {
          const data = await response.json();
          FetchData(data);
          console.log(url);
        }
      } catch (err) {
        console.log(err);
      }
    };
    getData();
  }, [url]);

  const handlePreviousPage = () => {
    const a = currentPage + 1;
    setCurrentPage(a);
  };

  const handleNextPage = () => {
    const a = currentPage - 1;
    setCurrentPage(a);
  };

  const handleStatus = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setStatus(event.target.value);
  };
  const handleGender = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setGender(event.target.value);
  };
  const handleSpecies = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setSpecies(event.target.value);
  };
  return (
    <>
      <SearchBar />
      {error && <h2 className="error">No Characters Found</h2>}
      {isPending && (
        <div className="loader">
          <Spinner />
        </div>
      )}
      <section className="card-section">
        <nav>
          <ul className="pagination">
            <li className="page-item">
              <button className="page-link" onClick={handlePreviousPage}>
                Next
              </button>
            </li>
            <li className="page-item">
              <span>
                {currentPage}/{pagesCount}
              </span>
            </li>
            <li className="page-item">
              <button className="page-link" onClick={handleNextPage}>
                Previous
              </button>
            </li>
            <li className="page-item">
              <span>Status</span>
              <select className="page-link" onChange={handleStatus}>
                <option value="Alive">Alive</option>
                <option value="Dead">Dead</option>
                <option value="Unknown">Unknown</option>
              </select>
            </li>
            <li className="page-item">
              <span>Species</span>
              <select className="page-link" onChange={handleSpecies}>
                {spec.map((ele, i) => {
                  return (
                    <option value={ele} key={i}>
                      {ele}
                    </option>
                  );
                })}
              </select>
            </li>
            <li className="page-item">
              <span>Genders</span>
              <select className="page-link" onChange={handleGender}>
                {genders.map((el, i) => {
                  return (
                    <option value={el} key={i}>
                      {el}
                    </option>
                  );
                })}
              </select>
            </li>
          </ul>
        </nav>
        {/* <nav> */}
        {/* <ul className="pagination">
            <li className="page-item">
              <button className="page-link" onClick={handleFPage}>
                20
              </button>
            </li>
            <li className="page-item">
              <button className="page-link" onClick={handleSPage}>
                30
              </button>
            </li>
            <li className="page-item">
              <button className="page-link" onClick={handleTPage}>
                40
              </button>
            </li>
          </ul>
        </nav> */}
        <div className="card-items">
          {character.map((char: Icharacter, i: any) => {
            const idx = i;
            return (
              <Link key={idx} to={`${char.id}`}>
                <CardItem key={idx} char={char} />
              </Link>
            );
          })}
        </div>
        <div className="pages">
          {pages.map((page: any, index: any) => {
            return (
              <span
                className={currentPage === page ? 'current-page' : 'page'}
                key={index}
                onClick={() => setCurrentPage(page)}
              >
                {page}
              </span>
            );
          })}
        </div>
      </section>
    </>
  );
};
