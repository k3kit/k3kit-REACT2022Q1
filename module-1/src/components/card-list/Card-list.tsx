import React from 'react';
import data from '../card/data';
import Card from '../card/Card';
import './style.css';

export const CardList = (): JSX.Element => {
  return (
    <ul className="card-item">
      {data.map((el) => {
        return (
          <Card
            key={el.id}
            avatar={el.avatar}
            first_name={el.first_name}
            gender={el.gender}
            email={el.email}
          />
        );
      })}
    </ul>
  );
};
