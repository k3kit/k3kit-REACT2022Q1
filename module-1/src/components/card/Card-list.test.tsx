import { render, screen } from '@testing-library/react';
import Card from './Card';
import { CardList } from '../card-list/Card-list';
import data from './data';

describe('test ', () => {
  test(' all card test', () => {
    render(<CardList />);
    const cards = screen.getAllByTestId('card');
    expect(cards).toHaveLength(data.length);
  });
  test('list snapshot', () => {
    const list = render(<CardList />);
    expect(list).toMatchSnapshot();
  });
});

describe('test ', () => {
  test(' one card test', () => {
    render(
      <Card
        avatar={'https://robohash.org/culpamagnameius.png?size=50x50&set=set1'}
        first_name={'abc'}
        email={'mail'}
        gender={'gen'}
      />
    );
    expect(screen.getByRole('img')).toBeInTheDocument();
    expect(screen.getByText(/abc/)).toBeInTheDocument();
    expect(screen.getByText(/mail/)).toBeInTheDocument();
    expect(screen.getByText(/gen/)).toBeInTheDocument();
    screen.debug();
  });
});
