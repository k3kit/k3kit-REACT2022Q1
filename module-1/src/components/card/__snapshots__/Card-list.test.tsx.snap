// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test  list snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="card-item"
      >
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/culpamagnameius.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Ignaz
            </p>
            <p
              class="email"
            >
              Email: 
              ibrasted0@arizona.edu
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/saepelaboriosamofficia.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Raoul
            </p>
            <p
              class="email"
            >
              Email: 
              rfissenden1@google.ru
            </p>
            <p
              class="gender"
            >
              Gender: 
              Female
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/autinest.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Mavis
            </p>
            <p
              class="email"
            >
              Email: 
              msimcock2@simplemachines.org
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/asperiorestemporaincidunt.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Ave
            </p>
            <p
              class="email"
            >
              Email: 
              abovingdon3@tmall.com
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/velitporroid.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Eolande
            </p>
            <p
              class="email"
            >
              Email: 
              efrascone4@wufoo.com
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/debitisomnisnesciunt.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Arlie
            </p>
            <p
              class="email"
            >
              Email: 
              alansly5@360.cn
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/temporaetmaiores.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Rodie
            </p>
            <p
              class="email"
            >
              Email: 
              rdoyley6@hubpages.com
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
        <li
          class="card"
          data-testid="card"
        >
          <img
            alt="avatar"
            class="card-img"
            src="https://robohash.org/quiaquidemmagni.png?size=50x50&set=set1"
          />
          <div
            class="card-description"
          >
            <p
              class="first-name"
            >
              Name:
              Tymon
            </p>
            <p
              class="email"
            >
              Email: 
              tmasterman7@scribd.com
            </p>
            <p
              class="gender"
            >
              Gender: 
              Male
            </p>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="card-item"
    >
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/culpamagnameius.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Ignaz
          </p>
          <p
            class="email"
          >
            Email: 
            ibrasted0@arizona.edu
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/saepelaboriosamofficia.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Raoul
          </p>
          <p
            class="email"
          >
            Email: 
            rfissenden1@google.ru
          </p>
          <p
            class="gender"
          >
            Gender: 
            Female
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/autinest.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Mavis
          </p>
          <p
            class="email"
          >
            Email: 
            msimcock2@simplemachines.org
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/asperiorestemporaincidunt.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Ave
          </p>
          <p
            class="email"
          >
            Email: 
            abovingdon3@tmall.com
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/velitporroid.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Eolande
          </p>
          <p
            class="email"
          >
            Email: 
            efrascone4@wufoo.com
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/debitisomnisnesciunt.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Arlie
          </p>
          <p
            class="email"
          >
            Email: 
            alansly5@360.cn
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/temporaetmaiores.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Rodie
          </p>
          <p
            class="email"
          >
            Email: 
            rdoyley6@hubpages.com
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
      <li
        class="card"
        data-testid="card"
      >
        <img
          alt="avatar"
          class="card-img"
          src="https://robohash.org/quiaquidemmagni.png?size=50x50&set=set1"
        />
        <div
          class="card-description"
        >
          <p
            class="first-name"
          >
            Name:
            Tymon
          </p>
          <p
            class="email"
          >
            Email: 
            tmasterman7@scribd.com
          </p>
          <p
            class="gender"
          >
            Gender: 
            Male
          </p>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
